### DISCUSSION TITLE
Maths 2.0

### DISCUSSION THREAD

**Discussion by MatiDragon-YT** [sannybuilder/dev#138] [Ideas]
How about these shortcuts?
```
int 0@, 1@, 2@ = 1
float 3@ = 6.5, 4@

// calc 1
0@ = 1@ - 2@ * 3@ / 4@

// calc 2
5@ += 6@(7@ + 8@)

// calc 3
9@ = 8@ + 7@(6@ * 5@)

// at func
($actor, 20@(21@ -= 2, 10i)) = CreateAtCar(#FAM1, #JESTER, 123.45, 0@(1@ = 4, 5i), 6@ += 2.27)

// deltaTime
float 0@ = 5 + deltaTime * 2
$962 = deltaTime * 3.5

// converts
float 1@ = 75.6 * 3.3
1@ = toInteger 2@
3@.toFloat 4@
```
To generate this code:
```
var
0@ : int
1@ : int
2@ : int = 1
3@ : float = 6.5
4@ : float
end

// calc 1
XXXX: 0@ =1@
XXXX: 0@ -= 2@
XXXX: 0@ *= 3@
XXXX: 0@ /= 4@

// calc 2
XXXX: 7@ += 8@
XXXX: 6@ = 7@
XXXX: 5@ += 6@

// calc 3
XXXX: 6@ *= 5@
XXXX: 7@ = 6@
XXXX: 8@ += 7@
XXXX: 9@ = 8@

// at func
XXXX: 21@ -= 2
XXXX: 1@ = 4
XXXX: 6@ += 2.27
XXXX: call_scm_func @CreateAtCar flag 5 params #FAM1 #JESTER 123.45 0@(1@, 5i) 6@ = $actor 20@(21@, 10i)

// deltaTime
XXXX: 0@ = 5
XXXX: 0@ += frame_delta_time * 2
XXXX: $962 = frame_delta_time * 3.5

// converts
XXXX: 1@ = 75.6
XXXX: 1@ *= 3.3
XXXX: 1@ = to_integer 2@
XXXX: 3@ = to_float 4@
```

**Comment by wmysterio** [comment]
Variable type declaration is nice.

`calc1`, `calc2` and `calc3` has a bizarre and not typical syntax in programming. The result is unexpected. 

Other look here #115.

I'm will planned create a plugin with opcodes:
```text
0@ = 4
1@ = 2
XXXX: calc_int $result = "1 + 2"             // $result = 1+2 => 3
XXXX: calc_int $result = "1 + %d / %d" 0@ 1@ // $result = 1+0@/1@ => 1+(4/2) => 3

0@ = 2.0
1@ = 3.0
XXXX: calc_float $result = "1.0 - 2.0"           // $result = 1.0-2.0 => -1.0
XXXX: calc_float $result = "1.5 * %f ^ %f" 0@ 1@ // $result = 1.5*0@^1@ => 1.5*(2.0^3.0) => 12.0
```
I wrote like this in СLEO-function, but the syntax was terrible. A plugin will do it better.

**Reply by MatiDragon-YT** [reply]
### Antiguamente...
> complex expressions are fine and easy to implement until we got to the point where we need to introduce a temporary variable to calculate the result. consider:
> `$var = ($a + $a) * ($a - $a)`
> this should unwrap to:
> `$temp1 = $a`
> `$temp1 += $a`
> `$temp2 = $a`
> `$temp2 -= $a`
> `$var = $temp1`
> `$var *= $temp2`
which is again possible to implement but requires extra memory managements. Due to limits on variables the compiler can't arbitrarily add more hidden variables (neither global, as it affects the main.scm layout by increasing varspace in the header, nor local as their amount is very low). Here is the dilemma, how to circumvent it.
<img alight="right" alt="By x87 at #58"/>

### Making that clear
- In `calc1` the calculations are made on the first variable (`0@`).  This means that `1@`, `2@` and `3@` will be set as the second parameter of the OP XXXX.
- In `calc2` we make the aforementioned accessible.  We define where we will store these temporary numbers, either in a local (`@`) or global (`$`) variable.  As in schools, we first calculate what is inside the parentheses `()` and then proceed with the others.
The syntax is simple `$result($var1 +{operator} $var2)`.
- In `calc3` a combination of `calc1` and `cacl2` is shown.

_Your idea is not bad either._ :)

**Reply by wmysterio** [reply]
Your example does not solve the dilemma that x87 spoke about. He just does other things :)

Not a bad idea, but I'd rather not use the `=` operator so that beginners and experienced programmers can learn the syntax faster.

Perhaps the lambda expression `=>` operator would be more appropriate.

```text
// calc 1
0@ => 1@ - 2@ * 3@ / 4@
XXXX: 0@ = 1@
XXXX: 0@ -= 2@
XXXX: 0@ *= 3@
XXXX: 0@ /= 4@
```

**Reply by MatiDragon-YT** [reply]
> Your example does not solve the dilemma that x87 spoke about. He just does other things :)

But in this way we would not be affecting the structure of the scm with the limits of the game... ;:/
